tests/digraph/test_substitute_node_with_subgraph.py

```diff
@@ -0,0 +1,150 @@
+
+    def test_node_filter(self):
+        in_graph = retworkx.PyDiGraph()
+        in_graph.add_node(0)
+        in_graph.add_child(0, 1, "edge")
+        res = self.graph.substitute_node_with_subgraph(
+            2,
+            in_graph,
+            lambda _, __, ___: 0,
+            node_filter=lambda node: node == 0,
+        )
+        self.assertEqual(
+            [(0, 1), (3, 4), (1, 5), (5, 3)], self.graph.edge_list()
+        )
+        self.assertEqual(res, {0: 5})
+
+    def test_edge_weight_modifier(self):
+        in_graph = retworkx.PyDiGraph()
+        in_graph.add_node(0)
+        in_graph.add_child(0, 1, "edge")
+        res = self.graph.substitute_node_with_subgraph(
+            2,
+            in_graph,
+            lambda _, __, ___: 0,
+            edge_weight_map=lambda edge: edge + "-migrated",
+        )
+        self.assertEqual(
+            [(0, 1), (3, 4), (5, 6), (1, 5), (5, 3)], self.graph.edge_list()
+        )
+        self.assertEqual("edge-migrated", self.graph.get_edge_data(5, 6))
+        self.assertEqual(res, {0: 5, 1: 6})
+
+    def test_none_mapping(self):
+        in_graph = retworkx.PyDiGraph()
+        in_graph.add_node(0)
+        in_graph.add_child(0, 1, "edge")
+        res = self.graph.substitute_node_with_subgraph(
+            2, in_graph, lambda _, __, ___: None
+        )
+        self.assertEqual([(0, 1), (3, 4), (5, 6)], self.graph.edge_list())
+        self.assertEqual(res, {0: 5, 1: 6})
+
+    def test_multiple_mapping(self):
+        graph = retworkx.generators.directed_star_graph(5)
+        in_graph = retworkx.generators.directed_star_graph(3, inward=True)
+
+        def map_function(source, target, _weight):
+            if target > 2:
+                return 2
+            return 1
+
+        res = graph.substitute_node_with_subgraph(0, in_graph, map_function)
+        self.assertEqual({0: 5, 1: 6, 2: 7}, res)
+        expected = [(6, 5), (7, 5), (7, 4), (7, 3), (6, 2), (6, 1)]
+        self.assertEqual(expected, graph.edge_list())
+
+    def test_multiple_mapping_full(self):
+        graph = retworkx.generators.directed_star_graph(5)
+        in_graph = retworkx.generators.directed_star_graph(
+            weights=list(range(3)), inward=True
+        )
+        in_graph.add_edge(1, 2, None)
+
+        def map_function(source, target, _weight):
+            if target > 2:
+                return 2
+            return 1
+
+        def filter_fn(node):
+            return node > 0
+
+        def map_weight(_):
+            return "migrated"
+
+        res = graph.substitute_node_with_subgraph(
+            0, in_graph, map_function, filter_fn, map_weight
+        )
+        self.assertEqual({1: 5, 2: 6}, res)
+        expected = [
+            (5, 6, "migrated"),
+            (6, 4, None),
+            (6, 3, None),
+            (5, 2, None),
+            (5, 1, None),
+        ]
+        self.assertEqual(expected, graph.weighted_edge_list())
+
```

Comment:
Maybe could add a case with where the substituted node have an inward and outward edge.

Reply:
Done in: https://github.com/Qiskit/retworkx/pull/312/commits/cca111b90121a94639911df788297429efa988b8